src/cmd/flux-simulator.py

```diff
@@ -0,0 +1,610 @@
+            return self._jobspec
+
+        assert self.ncpus % self.nnodes == 0
+        core = create_resource("core", self.ncpus / self.nnodes)
+        slot = create_slot("task", 1, [core])
+        if self.nnodes > 0:
+            resource_section = create_resource("node", self.nnodes, [slot])
+        else:
+            resource_section = slot
+
+        jobspec = {
+            "version": 1,
+            "resources": [resource_section],
+            "tasks": [
+                {
+                    "command": ["sleep", "0"],
+                    "slot": "task",
+                    "count": {"per_slot": 1},
+                }
+            ],
+            "attributes": {"system": {"duration": self.timelimit}},
+        }
+
+        self._jobspec = jobspec
+        return self._jobspec
+
+    def submit(self, flux_handle):
+        jobspec_json = json.dumps(self.jobspec)
+        logger.log(9, jobspec_json)
+        flags = 0
+        if logger.isEnabledFor(logging.DEBUG):
+            logger.debug("Submitting job with FLUX_JOB_DEBUG enabled")
+            flags = flux.constants.FLUX_JOB_DEBUG
+        self._submit_future = flux.job.submit_async(flux_handle, jobspec_json, flags=flags)
+
+    @property
+    def jobid(self):
+        if self._jobid is None:
+            if self._submit_future is None:
+                raise ValueError("Job was not submitted yet. No ID assigned.")
+            logger.log(9, "Waiting on jobid")
+            self._jobid = flux.job.submit_get_id(self._submit_future)
+            self._submit_future = None
+            logger.log(9, "Received jobid: {}".format(self._jobid))
+        return self._jobid
+
+    @property
+    def complete_time(self):
+        if self.start_time is None:
+            raise ValueError("Job has not started yet")
+        return self.start_time + self.elapsed_time
+
+    def start(self, flux_handle, start_msg, start_time):
+        self.start_time = start_time
+        self._start_msg = start_msg.copy()
+        flux_handle.respond(
+            self._start_msg, payload={"id": self.jobid, "type": "start", "data": {}}
+        )
+
+    def complete(self, flux_handle):
+        # TODO: emit "finish" event
+        flux_handle.respond(
+            self._start_msg,
+            payload={"id": self.jobid, "type": "finish", "data": {"status" : 0}}
+        )
+        # TODO: emit "done" event
+        flux_handle.respond(
+            self._start_msg,
+            payload={"id": self.jobid, "type": "release", "data": {"ranks" : "all", "final": True}}
+        )
+
+    def cancel(self, flux_handle):
+        flux.job.RAW.cancel(flux_handle, self.jobid, "Canceled by simulator")
+
+    def insert_apriori_events(self, simulation):
+        # TODO: add priority to `add_event` so that all submits for a given time
+        # can happen consecutively, followed by the waits for the jobids
+        simulation.add_event(self.submit_time, lambda: simulation.submit_job(self))
```

Comment:
Astute observation from @mrwyattii: this logic should be contained in the simulator along with the other job event additions.

I originally planned to have the Job add all of it's own events so that the Simulation could remain agnostic of the job's lifecycle (submit -> run -> complete).  That would make adding new job states like depend, grow, and shrink only require modifying the Job class.  It would also allow adding new entities like a Resource (e.g., node, filesystem) more modular; they would each handle their own event adding and the Simulation could remain ignorant of their lifecycles.  But that is probably left for another day and different PR.